packages/context/src/inject.ts

```diff
@@ -414,11 +420,20 @@ function resolveByFilter(
   injection: Readonly<Injection>,
   session?: ResolutionSession,
 ) {
-  const bindingFilter = injection.metadata!.bindingFilter;
+  const targetType = inspectTargetType(injection);
+  if (injection.metadata && injection.metadata.decorator === '@inject.view') {
+    if (targetType && targetType !== ContextView) {
+      const targetName = ResolutionSession.describeInjection(injection)!
+        .targetName;
+      throw new Error(
+        `The type of ${targetName} (${targetType.name}) is not ContextView`,
+      );
+    }
+  }
+  const bindingFilter = injection.bindingSelector as BindingFilter;
   const view = new ContextView(ctx, bindingFilter);
   const watch = injection.metadata!.watch;
 
-  const targetType = inspectTargetType(injection);
   if (targetType === Function) {
     if (watch !== false) view.watch();
     return view.asGetter();
```

Comment:
I thought that we agreed to get rid of this auto-magic determining what is being injected depending on the target type, and use the selector type to decide on what should be injected (a single key vs. a binding filter).

It is also not clear to me how is `@inject.getter()` and `@inject()` dealing with the difference between injecting a single value (by binding key) or injecting an array of values matching a filter - could you please point me in the right direction?

Ideally, I'd like us to remove `resolveByFilter` and replace it with smaller `resolve` functions specialized on different injections (resolve view, resolve array of values, resolve a getter of array of values).

Maybe I am missing some requirements and constraints that are explaining the current design?

Let's discuss how we can simplify this part of this implementation.

Reply:
Removed the magic and added meaningful errors.

Reply:
I still find the new solution as too magical, it's not clear to me how the different resolution types work when I read only this part of the codebase.

I guess I have to spend more time on reading not only `resolveByFilter` but also other related functions to better understand what's going on here.

Here is the problem though: such complexity makes it difficult to understand how injection works, people new to LB4 codebase (or even us 6-12 months down the line) will have to spent extra time and effort to familiarize with this part of our codebase.

Can you find such an internal design that will make the implementation easier to read, understand and reason about?

Reply:
Fixed.